Style: Harmonize header includes in platform ports

akien-mga opened the pull request:
This applies our [existing style guide](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#header-includes), and adds a new rule to that style guide for modular components such as platform ports and modules:

> Includes from the platform port or module should be included with relative paths (relative to the root folder of the modular component, e.g. `platform/linuxbsd/`), in their own section before Godot's "core" includes.

I'll add this to the [style guide](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#header-includes) if/when we reach consensus.

This helps differentiate between "local" code and the rest of the engine which can be treated as a library. The order of includes therefore flows from the most local (`header.h` for `header.cpp`) to the most external (`<string.h>`, `<windows.h>`, etc.). It's not perfect for platform code as we also have code in `drivers/` which has a similar kind of "locality" to headers in the platform folder, but rules can never be perfect ;)

This also helps (at least theoretically) with keeping module components modular. If we want to rename a platform or module folder, or more importantly, if people need to maintain a thirdparty platform or module, keeping the component's own files with relative includes help with the plug 'n play mechanism for `platform/` and `modules/`.

Same treatment applied to modules: #75944.

akien-mga commented on platform/ios/display_server_ios.mm:
```diff
@@ -31,17 +31,18 @@
 #import "keyboard_input_view.h"
-#include "os_ios.h"
-#include "tts_ios.h"
+#import "os_ios.h"
+#import "tts_ios.h"
```
@bruvzg In iOS code, we seem to use both `#include` and `#import` for the `platform/ios/*.h` includes. I tried to standardize on `#import` but if there was a good reason for keeping some using `#include`, I can change it back.

bruvzg replied:
`import` can be safely used without include guards in the header, Godot headers do have guards, so there's no difference.

akien-mga replied:
I suggest we standardize on using `#import` for our own headers which match `.m`/`.mm` files, even if some of them seem to be pure C/C++. Would that make sense?

So basically `#import` for any of those:
```
$ ls platform/ios/*.h
platform/ios/app_delegate.h    platform/ios/display_server_ios.h  platform/ios/godot_view_renderer.h  platform/ios/keyboard_input_view.h  platform/ios/os_ios.h           platform/ios/view_controller.h
platform/ios/device_metrics.h  platform/ios/godot_app_delegate.h  platform/ios/ios.h                  platform/ios/key_mapping_ios.h      platform/ios/platform_config.h  platform/ios/vulkan_context_ios.h
platform/ios/display_layer.h   platform/ios/godot_view.h          platform/ios/joypad_ios.h           platform/ios/logo_svg.gen.h         platform/ios/tts_ios.h
```

akien-mga replied:
Did that (including for `drivers/{coreaudio,coremidi}/`).

bruvzg commented on platform/ios/tts_ios.h:
```diff
@@ -31,18 +31,18 @@
 
-#include "core/string/ustring.h"
-#include "core/templates/list.h"
-#include "core/templates/rb_map.h"
```
> I don't know if that's intentional 

No, probably it was missing during conversion.

akien-mga replied:
I included the simple replacement in this PR, but I can move it to a dedicated one if needed.

akien-mga replied:
I guess this is why the replacement wasn't done ;)
```
 scons: *** [platform\windows\tts_windows.windows.editor.x86_64.obj] Error 2
Error: .\core/templates/hash_map.h(85): error C2668: 'HashMapHasherDefault::hash': ambiguous call to overloaded function
.\core/templates/hashfuncs.h(327): note: could be 'uint32_t HashMapHasherDefault::hash(const int8_t)'
.\core/templates/hashfuncs.h(326): note: or       'uint32_t HashMapHasherDefault::hash(const uint8_t)'
.\core/templates/hashfuncs.h(325): note: or       'uint32_t HashMapHasherDefault::hash(const int16_t)'
.\core/templates/hashfuncs.h(324): note: or       'uint32_t HashMapHasherDefault::hash(const uint16_t)'
.\core/templates/hashfuncs.h(323): note: or       'uint32_t HashMapHasherDefault::hash(const int32_t)'
.\core/templates/hashfuncs.h(322): note: or       'uint32_t HashMapHasherDefault::hash(const uint32_t)'
.\core/templates/hashfuncs.h(321): note: or       'uint32_t HashMapHasherDefault::hash(const double)'
.\core/templates/hashfuncs.h(320): note: or       'uint32_t HashMapHasherDefault::hash(const float)'
.\core/templates/hashfuncs.h(319): note: or       'uint32_t HashMapHasherDefault::hash(const int64_t)'
.\core/templates/hashfuncs.h(318): note: or       'uint32_t HashMapHasherDefault::hash(const uint64_t)'
.\core/templates/hashfuncs.h(311): note: or       'uint32_t HashMapHasherDefault::hash(const char32_t)'
.\core/templates/hashfuncs.h(310): note: or       'uint32_t HashMapHasherDefault::hash(const char16_t)'
.\core/templates/hashfuncs.h(309): note: or       'uint32_t HashMapHasherDefault::hash(const wchar_t)'
.\core/templates/hash_map.h(85): note: while trying to match the argument list '(const TKey)'
        with
        [
            TKey=ULONG
        ]
.\core/templates/hash_map.h(84): note: while compiling class template member function 'uint32_t HashMap<ULONG,TTS_Windows::UTData,HashMapHasherDefault,HashMapComparatorDefault<TKey>,DefaultTypedAllocator<HashMapElement<TKey,TValue>>>::_hash(const TKey &) const'
        with
        [
            TKey=ULONG,
            TValue=TTS_Windows::UTData
        ]
.\core/templates/hash_map.h(238): note: see reference to function template instantiation 'uint32_t HashMap<ULONG,TTS_Windows::UTData,HashMapHasherDefault,HashMapComparatorDefault<TKey>,DefaultTypedAllocator<HashMapElement<TKey,TValue>>>::_hash(const TKey &) const' being compiled
        with
        [
            TKey=ULONG,
            TValue=TTS_Windows::UTData
        ]
D:\a\godot\godot\platform\windows\tts_windows.h(57): note: see reference to class template instantiation 'HashMap<ULONG,TTS_Windows::UTData,HashMapHasherDefault,HashMapComparatorDefault<TKey>,DefaultTypedAllocator<HashMapElement<TKey,TValue>>>' being compiled
        with
        [
            TKey=ULONG,
            TValue=TTS_Windows::UTData
        ]
```

akien-mga replied:
Should be done with #75933 (untested).

akien-mga commented:
Marked for cherry-picking (if approved) to keep the codebases in sync. Should be relatively easy to cherry-pick / solve merge conflicts by redoing the changes on affected files.

RandomShaper commented on platform/windows/display_server_windows.cpp:
```diff
@@ -30,23 +30,23 @@
 #include "drivers/gles3/rasterizer_gles3.h"
 #endif
 
+#ifndef DWMWA_USE_IMMERSIVE_DARK_MODE
```
This looks like an addendum to dwmapi.h in case it's an old version. I think we can have a rule for these cases, but it's not trivial because I guess situations can vary a lot. In any case, I'd suggest the include and the addendum not having anything interleaved (besides maybe blank lines?).

akien-mga replied:
Fixed.

RandomShaper reviewed (commented):
Besides the comment I've added, looks good to me.

YuriSizov commented on platform/android/api/api.cpp:
```diff
@@ -30,10 +30,11 @@
 #include "api.h"
 
-#include "core/config/engine.h"
 #include "java_class_wrapper.h"
 #include "jni_singleton.h"
```
> Includes from the platform port or module should be included with relative paths (relative to the root folder of the modular component, e.g. `platform/linuxbsd/`)

Should this then be?..

```suggestion
#include "api/java_class_wrapper.h"
#include "api/jni_singleton.h"
```

akien-mga replied:
Indeed!

akien-mga replied:
Fixed.

YuriSizov commented on platform/android/display_server_android.cpp:
```diff
@@ -30,19 +30,23 @@
+#include "core/config/project_settings.h"
+
 #if defined(VULKAN_ENABLED)
+#include "vulkan/vulkan_context_android.h"
+
 #include "drivers/vulkan/rendering_device_vulkan.h"
-#include "platform/android/vulkan/vulkan_context_android.h"
 #include "servers/rendering/renderer_rd/renderer_compositor_rd.h"
 #endif
+
 #ifdef GLES3_ENABLED
 #include "drivers/gles3/rasterizer_gles3.h"
+
 #include <EGL/egl.h>
 #endif
```
Also should it be `#if defined(VULKAN_ENABLED)` and not `#ifdef VULKAN_ENABLED` here? Seems like a remnant of something where the condition was more complex.

akien-mga replied:
> Also should it actually be `#if defined(VULKAN_ENABLED)` and not `#ifdef VULKAN_ENABLED` here? Seems like a remnant of something where the condition was more complex.

We don't really have a standard for this. That's outside the scope of this PR IMO.

YuriSizov replied:
That's fair enough!

YuriSizov commented on platform/android/export/export_plugin.cpp:
```diff
@@ -30,6 +30,8 @@
 
 #include "export_plugin.h"
 
+#include "../logo_svg.gen.h"
+#include "../run_icon_svg.gen.h"
```
Doesn't look like paths in other places have to be relative. E.g. `platform/android/api/jni_singleton.h` including `jni_utils.h`.

Is there some conflict that prevents resolution?

akien-mga replied:
Technically both `#platform/android` and the current folder are in the include path, so both work indeed.

I've opted for using relative paths for the "local" includes, though I seem to have missed some. Maybe we could remove `#platform/<name>/` from the include path and force using paths relative to the current file?

akien-mga replied:
I rebased the PR and did the changes we discussed, so the include for e.g. `platform/android/export/export_plugin.h` in `platform/android/export/export.h` is done with `export/export_plugin.h`.

akien-mga replied:
So this didn't work in the end, because the `api` and `export` folders are special and compiled on all platforms for editor builds. We don't add all platform folders to the include path, so for the `api/*.cpp` and `export/*.cpp` to compile properly, they still need to use relative include paths.

YuriSizov replied:
Ah, shame. Well, at least most of the codebase is consistent after these changes. Platforms can be an exception

YuriSizov reviewed (approved):
I left some notes, but I support this convention in principle.

akien-mga committed:
Style: Harmonize header includes in platform ports

This applies our existing style guide, and adds a new rule to that style
guide for modular components such as platform ports and modules:

Includes from the platform port or module should be included with relative
paths (relative to the root folder of the modular component, e.g.
`platform/linuxbsd/`), in their own section before Godot's "core" includes.

The `api` and `export` subfolders also need to be handled as self-contained
(and thus use relative paths for their "local" includes) as they are all
compiled for each editor platform, without necessarily having the api/export
matching platform folder in the include path.
E.g. the Linux editor build will compile `platform/android/{api,export}/*.cpp`
and those need to use relative includes for it to work.
